src/lib/storage/frame_of_reference/frame_of_reference_encoder.hpp

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <limits>
+#include <memory>
+
+#include "storage/base_column_encoder.hpp"
+
+#include "storage/frame_of_reference_column.hpp"
+#include "storage/value_column.hpp"
+#include "storage/value_column/value_column_iterable.hpp"
+#include "storage/vector_compression/vector_compression.hpp"
+#include "types.hpp"
+#include "utils/enum_constant.hpp"
+
+namespace opossum {
+
+class FrameOfReferenceEncoder : public ColumnEncoder<FrameOfReferenceEncoder> {
+ public:
+  static constexpr auto _encoding_type = enum_c<EncodingType, EncodingType::FrameOfReference>;
+  static constexpr auto _uses_vector_compression = true;
+
+  template <typename T>
+  std::shared_ptr<BaseEncodedColumn> _on_encode(const std::shared_ptr<const ValueColumn<T>>& value_column) {
+    const auto alloc = value_column->values().get_allocator();
+
+    static constexpr auto block_size = FrameOfReferenceColumn<T>::block_size;
+
+    const auto size = value_column->size();
+    const auto num_blocks = size / block_size;
+
+    // holds the minimum of each block
+    auto block_minima = pmr_vector<T>{alloc};
+    block_minima.reserve(num_blocks);
+
+    // holds the uncompressed offset values
+    auto offset_values = pmr_vector<uint32_t>{alloc};
+    offset_values.reserve(size);
+
+    // holds whether a column value is null
+    auto null_values = pmr_vector<bool>{alloc};
+    null_values.reserve(size);
+
+    // used as optional input for the compression of the offset values
+    auto max_value = uint32_t{0u};
+
+    auto iterable = ValueColumnIterable<T>{*value_column};
```

Comment:
Type erasure wouldn’t make sense here anyway because the lambda is only instantiated for one type. No I did it to avoid unnecessary includes and `create_iterable_from_column.hpp` includes a lot.